src/modalities/loss_functions.py

```diff
@@ -41,3 +41,75 @@ def __call__(self, forward_batch: InferenceResultBatch) -> torch.Tensor:
         # Flatten the tokens
         loss = self.loss_fun(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
         return loss
+
+
+class NCELoss(torch.nn.Module):
+    def __init__(self, asymm: bool, temp: float):
+        """
+        This calculates the noise contrastive estimation loss between embeddings of two different modalities
+        Implementation slightly adapted from https://arxiv.org/pdf/1912.06430.pdf by adding symmeric and asymmetric loss
+
+        Args:
+            asymm (bool): specifies if the loss is calculated in one direction or both directions
+            temp (float): temprature value for regulating loss
+        """
+        super(NCELoss, self).__init__()
+        self.asymm = asymm
+        self.temp = temp
+
+    def forward(self, embedding1: torch.Tensor, embedding2: torch.Tensor, device: torch.device) -> torch.Tensor:
+        """
+        Args:
+            embedding1 (torch.Tensor): embeddings from modality 1 of size batch x dim
+            embedding2 (torch.Tensor): embeddings from modality 2 of size batch x dim
+            device (torch.device): torch device for calculating loss
+
+        Returns:
+            torch.Tensor: loss tensor
+        """
+        # calculating the similarity matrix of size (bxb)
+        sim_matrix = torch.matmul(embedding1, embedding2.t()) / self.temp
+        # numerator of loss: using similarity scores for all positive pairs (e.g., image and its caption)
+        numerator = sim_matrix * torch.eye(sim_matrix.shape[0], device=device)
+        numerator = numerator.sum(dim=0).view(sim_matrix.shape[0], -1)
+        numerator = torch.logsumexp(numerator, dim=1)
+        # denominator of loss: using all similarity scores for all pairs (positive and negative)
+        if self.asymm:
+            denominator = torch.logsumexp(sim_matrix, dim=1)
```

Comment:
I checked the paper and I am not sure if the two tests of positive and negative pairs, i.e., P_i and N_i, need to be of roughly equal size. Here, the positive examples would be largely underrepresented. Would this be an issue in practice?